mitiq/cdr/cdr_execution.py

```diff
@@ -0,0 +1,208 @@
+            - sigma_replace (float): Width of the Gaussian distribution used
+                                     for ``method_replace='gaussian'``.
+            - random_state (int): seed for sampling.
+
+            FITTING OPTIONS:
+
+            - include_intercept (Bool): whether or not to include an intercept
+                                        in the function mapping noisy to exact
+                                        expectation values. Defult is True as
+                                        in most cases it makes for a better
+                                        mitigation.
+    .. [Czarnik2020] : Piotr Czarnik, Andrew Arramsmith, Patrick Coles,
+        Lukasz Cincio, "Error mitigation with Clifford quantum circuit
+        data," (https://arxiv.org/abs/2005.10189).
+    .. [Lowe2020] : Angus Lowe, Max Hunter Gordon, Piotr Czarnik,
+        Andrew Arramsmith, Patrick Coles, Lukasz Cincio,
+        "Unified approach to data-driven error mitigation,"
+        (https://arxiv.org/abs/2011.01157)."""
+    # Extracting kwargs:
+    method_select = kwargs.get("method_select", "uniform")
+    method_replace = kwargs.get("method_replace", "closest")
+    random_state = kwargs.get("random_state", None)
+    intercept = kwargs.get("include_intercept", True)
+    training_set_generation_kwargs_keys = ["sigma_select", "sigma_replace"]
+    kwargs_for_training_set_generation = {
+        key: kwargs.get(key) for key in training_set_generation_kwargs_keys
+    }
+    training_circuits_list = generate_training_circuits(
+        circuit,
+        num_training_circuits,
+        fraction_non_clifford,
+        method_select,
+        method_replace,
+        random_state,
+        kwargs=kwargs_for_training_set_generation,
+    )
+    # If specified scale the noise in the circuit of interest and training
+    # circuits:
+    if scale_noise:
+        # identify number of noise levels:
+        if isinstance(noise_scaling_factors, float) or isinstance(
+            noise_scaling_factors, int
+        ):
+            noise_levels = 2
+        else:
+            noise_levels = len(noise_scaling_factors) + 1
+
+        training_circuits_list = scale_noise_in_circuits(
+            training_circuits_list, scale_noise, noise_scaling_factors
+        )
+        # redefine circuit as a list of circuits with scaled noise:
+        circuits = scale_noise_in_circuits(
+            circuit, scale_noise, noise_scaling_factors
+        )
+    else:
+        noise_levels = 1
+    
+    #print(len(circuits))
+    # run the training circuits in the desired backend and simulator:
+    results_dict_training_circuits = execute_training_circuits(
+        training_circuits_list,
+        executor,
+        simulator,
+        noise_levels,
+    )
+    # run the circuit of interest in the desired backend:
+    results_dict_circuit_of_interest = execute_circuit_of_interest(
+        circuits, executor
+    )
+
+    if isinstance(observable, list):
```

Comment:
Either (i) require it to always be a list or (ii) make it a list if it's not already one. We don't want to repeat this entire codeblock verbatim.